### PR TITLE
Make cors middleware compliant with the intended spec

### PR DESCRIPTION
**Pull request by alexandrebodin**
### What does it do?

Make cors middleware compliant with the intended spec

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
building exp to test real world usage: https://github.com/strapi/strapi/actions/runs/8575893524

**Comment by derrickmehaffy** [comment]
exp version: `0.0.0-experimental.00cd76085e4bb6c15688e9671f74f517f25926c1`

**Comment by derrickmehaffy** [comment]
Running a new one as I'm getting some odd sharp errors, develop might be broken or something :thinking: 

https://github.com/strapi/strapi/actions/runs/8576033020/job/23506249541

File: packages/core/strapi/package.json [hunk 1]
```diff
@@ -111,7 +111,7 @@
     "watch": "pack-up watch"
   },
   "dependencies": {
-    "@koa/cors": "3.4.3",
+    "@koa/cors": "5.0.0",
```
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/strapi/package.json:114]: they migrated to ts and drop some node version supports like we did in the past. 

Main change is the default origin which meant having to reconsider the way we did things on our side.

File: packages/core/strapi/src/middlewares/cors.ts [hunk 2]
```diff
@@ -46,17 +50,11 @@ export const cors: Common.MiddlewareFactory<Config> = (config) => {
         originList = origin;
       }
 
-      const whitelist = Array.isArray(originList) ? originList : originList.split(/\s*,\s*/);
-
-      const requestOrigin = ctx.headers.origin ?? '';
-      if (whitelist.includes('*')) {
-        return credentials ? requestOrigin : '*';
+      if (Array.isArray(originList)) {
```
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/strapi/src/middlewares/cors.ts:53]: explained in the general answer below

**Comment by derrickmehaffy** [comment]
New exp version: `0.0.0-experimental.d5b46d578a5c055b8dcc66939e1b5d540976fafb`

**Comment by alexandrebodin** [comment]
> @alexandrebodin I'm getting this on the experimental, idk if it has something to do with your changes but something is broken: image

nope not relateds, looks like a sharp binary issue

**Comment by derrickmehaffy** [comment]
updating branch and releasing a new exp to fix the sharp issue

**Comment by derrickmehaffy** [comment]
New exp with sharp fix (again): https://github.com/strapi/strapi/actions/runs/8634304417/job/23669419178

**Comment by derrickmehaffy** [comment]
@alexandrebodin as we discussed in Discord, we are setting the `Access-Control-Allow-Origin` to multiple values which isn't possible so best solution (I think) is to add the following:

- If our origin is an array, check the incoming `origin` header to see if it's in the array, if so set `Access-Control-Allow-Origin` to that origin
- If it's not in the array either we can just set `Access-Control-Allow-Origin` to the first value in the array or some non-sense domain (the only reason I would say the non-sense is we could specify the error message as the domain, not sure if there is a spec for that or not).

**Comment by derrickmehaffy** [comment]
New exp: https://github.com/strapi/strapi/actions/runs/8696106150

**Review by derrickmehaffy — APPROVED** [review]
Tested pretty every option related to the middleware origin config and it all looks good :+1:

**Comment by derrickmehaffy** [comment]
final exp: `0.0.0-experimental.469f8c0fe2f5ae4a405ec89049706cd7debab0f7`

**Review by Convly — APPROVED** [review]

**Comment by Pablo-Aldana** [comment]
This changes seems to break previous cors configuration and I couldn't find documentation about how to upgrade to the new changes.

Right now I have my CORS like this:

```
{
    name: 'strapi::cors',
    config: {
      //enabled: true,
      headers: '*',
      origin: env('CORS'),
    },
  },
  ```

And in the env(CORS) like this but with PROD values:
CORS=http://localhost:1338, http://localhost:8000, http://192.168.68.114:8000, http://192.168.68.114:3000, http://localhost:3000, http://192.168.68.114:1338

As soon as I update to this the strapi version 4.24.0 (https://github.com/strapi/strapi/releases/tag/v4.24.0)or higher, the web stop working due to cors issues, even if the URL is in the cors.

Screenshot_2024-06-13 23 40 57_studUG


Can anyone point me to the documentation or  update guide for this change?
I can create a new issue if needed

Many thanks

**Comment by alexandrebodin** [comment]
With the cors fix we had to do for security we removed the string list support as we don't know if it makes sense to process it this way or not. to fix this for your usecase you should split the string yourself 

`env('CORS').split(', ')`

**Pull request merged**